### PR TITLE
Ability to download loaders from remote hub

### PR DESCRIPTION
**Pull request by emptycrown**
Download and use loaders from the new loader hub. Super simple usage:

```
from gpt_index import download_loader

GoogleDocsReader = download_loader('GoogleDocsReader')
reader = GoogleDocsReader()
reader.load_data(document_ids=['1wf-y2pd9C878Oh-FmLH7Q_BQkljdm6TQal-c1pUfrec'])
```

### CODE REVIEW

File: gpt_index/readers/download.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+from langchain.utilities import RequestsWrapper
+
+from gpt_index.readers.base import BaseReader
+
+LOADER_HUB_URL = (
+    "https://raw.githubusercontent.com/emptycrown/loader-hub/main/loader_hub"
+)
+
+
+def download_loader(loaderClassName: str) -> BaseReader:
```
Comment (jerryjliu) [inline gpt_index/readers/download.py:18]: nit: arg name should be snake case e.g. loader_class

File: gpt_index/readers/download.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        A Loader.
+    """
+    requests = RequestsWrapper()
+    response = requests.run(f"{LOADER_HUB_URL}/library.json")
+    library = json.loads(response)
+
+    # Look up the loader id (e.g. `web/simple_web`)
+    loader_id = library[loaderClassName]["id"]
+    dirpath = ".modules"
+    loader_filename = loader_id.replace("/", "-")
```
Comment (jerryjliu) [inline gpt_index/readers/download.py:34]: ooc why do this? why not preserve the existing directory structure?

File: gpt_index/readers/download.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        # Create a new directory because it does not exist
+        os.makedirs(dirpath)
+
+    if not os.path.exists(loader_path):
+        response = requests.run(f"{LOADER_HUB_URL}/{loader_id}/base.py")
+        with open(loader_path, "w") as f:
+            f.write(response)
+
+    if not os.path.exists(requirements_path):
+        try:
```
Comment (jerryjliu) [inline gpt_index/readers/download.py:48]: if this is a matter of checking if it exists i'd more explicitly do  (cp'ed from stack overflow)
```
import requests
response = requests.get('http://www.example.com')
if response.status_code == 200:
    print('Web site exists')
else:
    print('Web site does not exist') 
```
a catchall for all exceptions is risky

File: gpt_index/readers/download.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+"""Download loader from the Loader Hub."""
+
+import json
+import os
+import subprocess
+import sys
+from importlib import util
+
+from langchain.utilities import RequestsWrapper
```
Comment (jerryjliu) [inline gpt_index/readers/download.py:9]: if you want to check the status code (see below), i'd prob just use the raw requests module

**Review by jerryjliu — COMMENTED** [review]
generally lgtm, some comments

File: gpt_index/readers/download.py (hunk 1 above)
Comment (emptycrown) [reply gpt_index/readers/download.py:18]: true -- switching context between js and python had weird effect on your brain

File: gpt_index/readers/download.py (hunk 2 above)
Comment (emptycrown) [reply gpt_index/readers/download.py:34]: Felt easier to not have to recursively create directories -- don't feel strongly about this though

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**